### PR TITLE
Update: Add `always`/`never` option to `eol-last` (fixes #6938)

### LINKED ISSUE #6938: eol-last: Add "none" option to enforce no line break at end of file
**What version of ESLint are you using?**
3.2.2
**What parser (default, Babel-ESLint, etc.) are you using?**
Babel-ESLint 6.1.2
**Please show your full configuration:**
<details>
<summary>.eslintrc</summary>
```json
{
	"env": {
		"browser": true,
		"es6": true,
		"node": true
	},
	"extends": [
		"eslint:recommended",
		"plugin:react/recommended"
	],
	"installedESLint": true,
	"parser": "babel-eslint",
	"parserOptions": {
		"ecmaFeatures": {
			"experimentalObjectRestSpread": true,
			"jsx": true
		},
		"sourceType": "module"
	},
	"plugins": [
		"filenames",
		"react"
	],
	"rules": {
		"accessor-pairs": [
			"error"
		],
		"array-bracket-spacing": [
			"error",
			"never"
		],
		"array-callback-return": [
			"error"
		],
		"block-scoped-var": [
			"error"
		],
		"brace-style": [
			"warn",
			"stroustrup"
		],
		"camelcase": [
			"error", {
				"properties": "always"
			}
		],
		"comma-dangle": [
			"warn",
			"never"
		],
		"comma-style": [
			"warn",
			"last"
		],
		"comma-spacing": [
			"warn", {
				"before": false,
				"after": true
			}
		],
		"complexity": [
			"warn"
		],
		"consistent-return": [
			"error"
		],
		"consistent-this": [
			"error"
		],
		"curly": [
			"error"
		],
		"default-case": [
			"error"
		],
		"dot-notation": [
			"error"
		],
		"dot-location": [
			"error",
			"property"
		],
		"eqeqeq": [
			"error"
		],
		"filenames/match-regex": [
			"error",
			"^[a-zA-Z.]+$"
		],
		"guard-for-in": [
			"error"
		],
		"indent": [
			"warn",
			"tab", {
				"SwitchCase": 1
			}
		],
		"jsx-quotes": [
			"warn",
			"prefer-double"
		],
		"key-spacing": [
			"error", {
				"beforeColon": false,
				"afterColon": true,
				"mode": "strict"
			}
		],
		"keyword-spacing": [
			"warn", {
				"before": true,
				"after": true
			}
		],
		"linebreak-style": [
			"error",
			"unix"
		],
		"max-statements-per-line": [
			"warn", {
				"max": 1
			}
		],
		"new-parens": [
			"warn"
		],
		"no-caller": [
			"error"
		],
		"no-case-declarations": [
			"error"
		],
		"no-cond-assign": [
			"error"
		],
		"no-console": [
			"off"
		],
		"no-dupe-args": [
			"error"
		],
		"no-dupe-keys": [
			"error"
		],
		"no-duplicate-case": [
			"error"
		],
		"no-empty": [
			"warn"
		],
		"no-empty-character-class": [
			"error"
		],
		"no-empty-function": [
			"warn"
		],
		"no-empty-pattern": [
			"error"
		],
		"no-eval": [
			"error"
		],
		"no-ex-assign": [
			"error"
		],
		"no-extend-native": [
			"error"
		],
		"no-extra-bind": [
			"error"
		],
		"no-extra-boolean-cast": [
			"warn"
		],
		"no-extra-semi": [
			"error"
		],
		"no-fallthrough": [
			"error"
		],
		"no-floating-decimal": [
			"error"
		],
		"no-func-assign": [
			"error"
		],
		"no-inline-comments": [
			"warn"
		],
		"no-invalid-regexp": [
			"error"
		],
		"no-implicit-coercion": [
			"warn"
		],
		"no-implicit-globals": [
			"error"
		],
		"no-implied-eval": [
			"error"
		],
		"no-iterator": [
			"error"
		],
		"no-labels": [
			"error"
		],
		"no-lone-blocks": [
			"warn"
		],
		"no-multiple-empty-lines": [
			"warn", {
				"max": 0,
				"maxBOF": 0,
				"maxEOF": 0
			}
		],
		"no-negated-in-lhs": [
			"error"
		],
		"no-multi-spaces": [
			"error"
		],
		"no-multi-str": [
			"error"
		],
		"no-native-reassign": [
			"error"
		],
		"no-new-func": [
			"error"
		],
		"no-new-wrappers": [
			"error"
		],
		"no-obj-calls": [
			"error"
		],
		"no-octal": [
			"error"
		],
		"no-octal-escape": [
			"error"
		],
		"no-param-reassign": [
			"warn"
		],
		"no-proto": [
			"error"
		],
		"no-redeclare": [
			"error", {
				"builtinGlobals": true
			}
		],
		"no-regex-spaces": [
			"error"
		],
		"no-return-assign": [
			"error"
		],
		"no-script-url": [
			"error"
		],
		"no-self-assign": [
			"error"
		],
		"no-self-compare": [
			"error"
		],
		"no-sequences": [
			"warn"
		],
		"no-sparse-arrays": [
			"error"
		],
		"no-this-before-super": [
			"error"
		],
		"no-throw-literal": [
			"error"
		],
		"no-trailing-spaces": [
			"warn"
		],
		"no-underscore-dangle": [
			"warn", {
				"allow": ["_rawDBType"]
			}
		],
		"no-unneeded-ternary": [
			"warn"
		],
		"no-delete-var": [
			"error"
		],
		"no-unexpected-multiline": [
			"error"
		],
		"no-unmodified-loop-condition": [
			"error"
		],
		"no-unreachable": [
			"error"
		],
		"no-unused-expressions": [
			"error"
		],
		"no-useless-call": [
			"error"
		],
		"no-useless-concat": [
			"error"
		],
		"no-useless-escape": [
			"error"
		],
		"no-void": [
			"error"
		],
		"no-warning-comments": [
			"warn", {
				"terms": [
					"todo",
					"fixme",
					"fix",
					"bugfix",
					"warn"
				],
				"location": "anywhere"
			}
		],
		"no-with": [
			"error"
		],
		"no-var": [
			"error"
		],
		"object-curly-spacing": [
			"error",
			"always"
		],
		"object-property-newline": [
			"error"
		],
		"object-shorthand": [
			"error"
		],
		"prefer-arrow-callback": [
			"warn"
		],
		"prefer-const": [
			"warn"
		],
		"prefer-reflect": [
			"warn"
		],
		"prefer-rest-params": [
			"warn"
		],
		"prefer-spread": [
			"warn"
		],
		"prefer-template": [
			"warn"
		],
		"quotes": [
			"warn",
			"double", {
				"allowTemplateLiterals": true
			}
		],
		"radix": [
			"error",
			"as-needed"
		],
		"react/display-name": [
			"off"
		],
		"require-jsdoc": [
			"warn", {
				"require": {
					"FunctionDeclaration": true,
					"MethodDefinition": true,
					"ClassDeclaration": true
				}
			}
		],
		"semi": [
			"error",
			"always"
		],
		"space-in-parens": [
			"warn",
			"never"
		],
		"space-infix-ops": [
			"warn"
		],
		"space-unary-ops": [
			"warn"
		],
		"strict": [
			"error",
			"never"
		],
		"template-curly-spacing": [
			"error",
			"never"
		],
		"use-isnan": [
			"error"
		],
		"valid-jsdoc": [
			"error", {
				"prefer": {
					"return": "return"
				},
				"requireReturn": false,
				"requireParamDescription": true,
				"requireReturnDescription": false,
				"requireReturnType": true,
				"preferType": {
					"String": "string",
					"Number": "number",
					"Object": "object",
					"Symbol": "symbol",
					"Function": "function",
					"Boolean": "boolean"
				}
			}
		],
		"yield-star-spacing": [
			"error",
			"after"
		],
		"yoda": [
			"error",
			"never"
		]
	}
}
```
</details>
**What did you do? Please include the actual source code causing the issue.**
I am trying to lint a file that ends in a single newline character, making contributors aware that they should configure their editors' saving configuration properly, as many of them append an invisible newline on file saves.
**What did you expect to happen?**
If the file ends in a single newline character, this should be a linting error.
**What actually happened? Please include the actual, raw output from ESLint.**
ESLint doesn't output any linting error, but it should.

### PR DESCRIPTION
**Pull request by kdex**
In short, `eol-last`'s behavior could thus far only enforce that files ended with an OS-specific EOL. The rule's negation (i.e. that files should not end with an EOL) wasn't possible.

This PR adds a `"none"` option to `eol-last` which adds a way to check that a file doesn't end with an EOL, which fixes issue #6938. Additionally, there's also a fixer and some tests.

Here's a short overview of changes:
- Replaced ternary check of options with object access
- Initialized `location` with both `column` and `line` directly (note that `line` is now set by `sourceCode.lines.length` instead of performing a `.split("\n")`)
- Divided `checkBadEOF`'s code flow into two paths, where one path treats different operating systems' EOLs just like before, while the other path treats enforcing their absence
- Added a fixer that removes EOLs from the end of the file
- Added some tests for `"none"` option
- Extended `eol-last`'s documentation to mention `"none"`

### CODE REVIEW

**Comment by kdex** [comment]
(The CLA error shouldn't be occurring; I've just signed it. Can I re-toggle it? Force-pushing doesn't help.)

**Comment by platinumazure** [comment]
The CLA check doesn't like single names (kdex in your case). Some sort of "manual approval" on our end is theoretically possible (though I don't have access to do that), but your best bet is to change your git config to have something akin to a first and last name configured, amend and force-push the commit, and re-sign the CLA with the new name. Sorry for the inconvenience!

**Comment by kdex** [comment]
@platinumazure Ah, thanks for the explanation! I'd really like to keep a global configuration for GitHub and work-related/private git projects; having multiple configurations set up would add unnecessary complexity to my backups (and break at other places, if I changed it globally). Is there someone I could ping for manual approval, possibly? I apologize!

**Comment by platinumazure** [comment]
@eslint/eslint-tsc I'm not really sure what our CLA policy is really supposed to be here. The goal, as I understand it, is to ensure all code contributions, represented by commits, are owned by the contributors and freely licensed to us. We use name and email matching to trace "who provided the contribution" and match against the CLA list. Is "having two or more names" really an important part of this check? Do we, perhaps, expect a real name (which isn't the same as two or more names) in case we need to re-verify that the contribution was licensed according to our terms (or do other follow-up)? The jQuery Foundation has its reasons for doing these things- what about us, do we have to follow their norms 100% since we're now part of the Foundation?

**Comment by pmcelhaney** [comment]
@kdex Can you not keep your global config and change it just for this project?

http://www.codeography.com/2011/08/05/project-specific-git-author.html

Otherwise, you should be able to override the author for just this commit.

```sh
git commit --amend --author="Your Name <your@email.com>"
```

File: lib/rules/eol-last.js [hunk 1]
```diff
@@ -34,23 +34,49 @@ module.exports = {
...
-                        message: "Newline required at end of file but not found.",
+                        message: "Newline not allowed at end of file.",
                         fix(fixer) {
-                            return fixer.insertTextAfterRange([0, src.length], linebreak);
+                            const finalEOLs = /(?:\r?\n)+$/,
+                                match = finalEOLs.exec(sourceCode.text),
+                                start = match.index,
+                                end = sourceCode.text.length;
+
+                            return fixer.replaceTextRange([start, end], "");
```
Comment (pmcelhaney) [inline lib/rules/eol-last.js]: Do we have tests for the fixer?

**Comment by kdex** [comment]
@pmcelhaney If I were to change the git configuration just for this project, I'd have to back it up, or else I'd be facing the exact same problem in case of future PRs (or even any PR that I'm creating from another machine, really). That's what I was trying to get at when I was referring to it as "unnecessary complexity" on my side.

Again, I do apologize for making this more difficult than it needs to be, but I don't understand the motivation for matching git usernames against a certain RegExp, given that they doesn't even need to match real names.

About the fixer: Please correct me if I'm wrong, but I assumed that the fixer output corresponds to the `output` property in the `invalid` array in `tests/lib/rules/eol-last.js`, so I would think that the answer is "yes".

**Comment by pmcelhaney** [comment]
I guess I don't understand what the unnecessary complexity is. If you sent the author configuration, what problem are you expecting to face next time?

(Nevermind the comment about tests. I figured it out about 10 seconds after I sent that comment.)

**Comment by kdex** [comment]
@pmcelhaney Here's what I'd expect to happen:

-  If I were to make a PR on another machine, I'd just get another CLA error, since my git configuration would only have been configured locally
- Now I'm in a situation where I'd have to setup n different machines. If I don't remember the name that I used to match @jquerybot's CLA RegExp, I'd have to `$ git log | grep kdex` on that machine just to find out about the name I crafted for the CLA so I can re-configure git. Rinse and repeat for every time I format my hard drive.
- If, instead of grepping, I were to craft myself a new name every time and re-sign the CLA, that'd make the commit tree weird and inconsistent

Note that the CLA check can't check for real names anyway, making the necessity for the name to contain a space incomprehensible to me. Dropping this rule, everyone could commit from anywhere without the need for project-specific setups.

**Comment by platinumazure** [comment]
@kdex I agree with you in spirit, if the only purpose of the CLA is to make sure that the same user who submitted commits has signed the CLA. That's why I've raised this to our TSC team. Please let's let them weigh this carefully and come to an informed decision.

**Comment by gyandeeps** [comment]
@kdex Side note, I also manage 2 repos on my machine (work and public).

I have this bash function which get executed on startup which sets the parameters correctly for my work repos:

```sh

# inside .bashrc file
updateGitInfo() {
    originUrl=$(git ls-remote --get-url origin)
    #re='work github url match regex'
    
    if [[ -n $originUrl && $originUrl =~ $re ]]
    then
        echo Work git
        #git config user.email "your_email@example.com"
        #git config user.name "Awesome dude"
    fi    
}

updateGitInfo
```

My global info always has public github information.

**Comment by vitorbal** [comment]
I think the CLA issue discussed here should have been fixed by the jQuery folks already. @platinumazure do you know if there's any way to force the bot to re-check the CLA?

**Comment by platinumazure** [comment]
Just force-push, I think.

**Comment by kdex** [comment]
Huh, seems like it still doesn't pass the CLA check. Are you sure that the change is live? [The signature page](https://contribute.jquery.org/CLA/) still mentions:

> The **name and email address** used to sign the CLA must match the name and email address in your git config.

I've also tried to re-sign the CLA, since I wasn't sure whether the back end just silently rejected all signatures without a space in the name, but to no avail.

**Comment by vitorbal** [comment]
@dmethvin @kborchers hey all, sorry to bother you with this, the CLA check for @kdex still says "Manual verification required", even after forcing a re-check. Shouldn't the license check allow users with a single name now after https://github.com/jquery/jquery-license/pull/70?

**Comment by kborchers** [comment]
My guess would be that the update didn't deploy due to the previous CLA disk issue. We will look into it ASAP.

**Comment by kborchers** [comment]
OK, so it looks like the changes made in jquery/jquery-license#70 did not remove the 2 name check. That is done elsewhere. We'll have to get that change implemented but in the meantime I did the manual verification for @kdex and this is all good now.

**Comment by kborchers** [comment]
I'll apologize for the noise in advance but we just pushed the real fix to the issue so I am going to remove the manual verification and recheck this PR. If it fails the CLA check I will re-validate the CLA signature and approve so you can merge and will test further elsewhere.

**Comment by kborchers** [comment]
I think we're good. Sorry again for the noise and thanks!

**Comment by platinumazure** [comment]
Thanks again for your help, @kborchers!

**Comment by vitorbal** [comment]
Thanks a lot, @kborchers! :heart:

**Comment by kborchers** [comment]
No problem :blush:

File: tests/lib/rules/eol-last.js [hunk 2]
```diff
@@ -54,6 +54,24 @@ ruleTester.run("eol-last", rule, {
...
+            code: "var a = 123;\r\n",
+            options: ["none"],
+            errors: [{ message: "Newline not allowed at end of file.", type: "Program" }],
+            output: "var a = 123;"
+        },
+        {
+            code: "var a = 123;\r\n\r\n",
+            options: ["none"],
+            errors: [{ message: "Newline not allowed at end of file.", type: "Program" }],
+            output: "var a = 123;"
```
Comment (vitorbal) [inline tests/lib/rules/eol-last.js]: Could you add some tests for when there are other newlines in the code that are not EOF newlines?

File: tests/lib/rules/eol-last.js [hunk 3]
```diff
@@ -54,6 +54,24 @@ ruleTester.run("eol-last", rule, {
             options: ["windows"],
             errors: [{ message: "Newline required at end of file but not found.", type: "Program" }],
             output: "var a = 123;\r\n   \r\n"
+        },
```
Comment (vitorbal) [inline tests/lib/rules/eol-last.js]: Thanks for adding these! Could you add a couple of new tests for the "valid" scenarios as well?

File: docs/rules/eol-last.md [hunk 4]
```diff
@@ -8,7 +8,8 @@ as output files to the terminal without interfering with shell prompts.
 
 ## Rule Details
 
-This rule requires at least one newline at the end of non-empty files.
+This rule enforces at least one newline (or absence thereof) at the end
```
Comment (vitorbal) [inline docs/rules/eol-last.md]: Could you update the title as well? "Require or disallow file to end with single newline"

File: lib/rules/eol-last.js [hunk 5]
```diff
@@ -20,7 +20,7 @@ module.exports = {
 
         schema: [
             {
-                enum: ["unix", "windows"]
```
Comment (vitorbal) [inline lib/rules/eol-last.js]: Could you update the meta.docs.description to reflect the new option?

File: docs/rules/eol-last.md [hunk 6]
```diff
@@ -42,3 +42,4 @@ This rule has a string option:
 
 * `"unix"` (default) enforces line feed (LF) as newline
 * `"windows"` enforces carriage return line feed (CRLF) as newline
+* `"none"` enforces that the last line ends with neither LF nor CRLF as newline
```
Comment (vitorbal) [inline docs/rules/eol-last.md]: I think this might read better for non-natives if it says "enforces that the last line does not end with a LF or CRLF as newline". In fact, I'm thinking maybe we could get rid of the "LF or CLRF" part altogether and just say: "enforces that the last line does not end with a newline". What do you think?

**Comment by vitorbal** [comment]
@kdex thanks for bearing with us! I think this is already very good, just had a few nitpicks, a couple more updates for documentation and a suggestion for a few more tests. Thanks for your work so far!

**Comment by kdex** [comment]
@vitorbal Thanks a lot for reviewing! I've updated the files accordingly and added five additional tests. Let me know if there's anything else that should be changed.

**Comment by nzakas** [comment]
Should "none" be "never" to match our other rules?

**Comment by platinumazure** [comment]
@nzakas I had a discussion with @kaicataldo about that. My reasoning (I favored "none") was that in this case, the single option described a newline style and needed to be an adjective, so "none" made more sense; whereas other cases are describing enforcement rate, and so needed to be an adverb (always/never).

Short version: Unix/windows/none felt like a more natural set to me than unix/windows/never.

I'm open to being persuaded otherwise, but at least @kaicataldo had agreed with me at one point.

**Comment by vitorbal** [comment]
@kdex thanks! LGTM from my part.

**Comment by kaicataldo** [comment]
I like "none" because, in the greater context of our rules, the presence of a "never" option seems to imply an "always" option to me (which was part of the discussion @platinumazure and I had), which this rule doesn't have.

Revisiting [this comment](https://github.com/eslint/eslint/issues/6938#issuecomment-242821557), I actually think if we want the greatest clarity and greatest consistency with our other rules we would want the options to be:
```js
    "no-multiple-empty-lines": ["error", "always"] // defaulting to "{ "style: "unix" }
    "no-multiple-empty-lines": ["error", "always", { "style": "windows" }]
    "no-multiple-empty-lines": ["error", "never"]
```

This could be done in a backwards compatible way.

**Comment by kdex** [comment]
I agree with @kaicataldo. It'd be nicer if we could use this opportunity to make `eol-last` behave like most other rules.

Also, I noticed that `eol-last`'s currently shipped implementation is faulty when it comes to CRLF. More specifically, [these four tests](https://github.com/eslint/eslint/blob/ca3d448802d4e8a31d1a881226c0dcb65bf24971/tests/lib/rules/eol-last.js#L29-L32) are wrong, since they treat a Windows EOL just like a Unix EOL; all of these cases should therefore be invalid, as per default, Unix EOLs (LF) are expected.

My last commit fixes this behavior and therefore [adds two new cases](https://github.com/kdex/eslint/commit/c3500b038b1a574a818c6077250cdf5f7c7695c5#diff-ee4595dededf876c50a08c5e1f1a9ca7R92): The case where a CRLF was expected, but an LF was seen, and the case where an LF was expected, but a CRLF was seen. These new cases, again, come with tests and fixers.

Moreover, I also implemented @kaicataldo's suggestions in a backwards-compatible way, adding lots of [new tests](https://github.com/kdex/eslint/commit/c3500b038b1a574a818c6077250cdf5f7c7695c5#diff-9d43dd4092b307d22db09ba20b3645dd) that test both the old and new way of specifying options. Hence, I've removed the support for `"none"` and migrated the code over to `"always"`/`"never"`.

**Comment by kdex** [comment]
I hope I'm not overhastily jumping to conclusions here. To me, it seemed that the general consensus was that "never" implies the existence of "always", which should have been implemented to begin with.

The only argument against `"always"`/`"never"` I can think of would be that the tests have become unnecessarily repetitive due to backwards-compatibility, which can be resolved by
- deprecating the old schema, and eventually removing support for it at some point
- re-introducing the `"windows"`, `"unix"`, `"none"` schema, accepting the inconsistency
- just being okay with tests being thorough :smile: 

Please let me know what the best approach would be here.

**Comment by kaicataldo** [comment]
@kdex We need to make sure the other team members are on board - I'm sorry if I didn't make that clear. Let's see what the others have to say, but thanks for all your amazing work on this regardless!

**Comment by pmcelhaney** [comment]
You're trying way too hard to shoehorn the always / never theme here. The question is what character(s) should go at the end of a file. Possible answers are "\r\n" (UNIX), "\n" (Windows), or "" (none, nothing, bare, empty, something along those lines). 

The rule makes perfect sense as it stands now. There's no need to change the semantics and introduce an eventual breaking change.

**Comment by mysticatea** [comment]
I like `"always"`/`"never"` approach. And I think the `style` additional option is extra. It's overlapped with `linebreak-style` rule, IMO.

**Comment by pmcelhaney** [comment]
Okay, if style is covered by another rule, that makes sense.

**Comment by kaicataldo** [comment]
I'm :+1: for @mysticatea's suggestion. The current iteration of the rule only uses the `unix`/`windows` option for auto-fixing. The check itself is only for [if the last character of the source code is a `\n`](https://github.com/eslint/eslint/blob/master/lib/rules/eol-last.js#L44), meaning that deprecating `unix`/`windows` in favor of `always`/`never` would only affect the auto-fix (which is already covered by the `linebreak-style` auto-fix). We can ensure backwards compatibility by keeping the `unix`/`windows` options and making them enable the `always` option.

To sum it up: could we just add `always`/`never` options (that require or disallow a `\n` at the end of the source code and then in the docs deprecate `unix`/`windows` and suggest using the new `always`/`never` options in conjunction with `linebreak-style`?

**Comment by kaicataldo** [comment]
@kdex Thanks for your patience. It looks like we have support [for this suggestion](https://github.com/eslint/eslint/pull/6952#issuecomment-244768743). Does that make sense? I think we should leave the current fixing behavior if either of the old options are used to ensure complete backwards compatibility. If the new options are used, the user can rely on `linebreak-style` to do the autofixing. 

Please ping back if I can help clarify anything :) Again, appreciate your work on this!

**Comment by kdex** [comment]
@kaicataldo The current implementation in the PR should almost be equivalent to what you suggested. One difference, though, would be that `eol-last` itself expects an object with a `style` property where the user would pass his desired EOL flavor.

If I understand you correctly, I should remove this object and just assume that the newline flavor is `\n`, which will be fixed when used in conjunction with `linebreak-style`. (And of course, I should then update the documentation accordingly.) Is this correct?

**Comment by kaicataldo** [comment]
Correct, the schema should just be:
```js
schema: [
  {
    enum: ["always", "never", "unix", "windows"]
  }
]
```

`"always"` and `"newline"` will always check for and require/disallow a `\n` (Unix-style linebreak) and autofix by adding or removing a `\n` character. 

If `"unix"` is specified, it will do the same check and autofix as `"always"`.

If `"windows"` is specified, it will do the same check as `"always"`  for a `\n` character, but autofix with the Windows-style linebreak of `\r\n` (for backwards compatibility).

**Comment by kdex** [comment]
Thanks a lot for the clarification, @kaicataldo! :+1:

I've adjusted the schema to your description and was able to remove a lot of code/tests, relying on the fact that `linebreak-style` will cover us here. I've also updated the documentation accordingly.

Please let me know if I need to change anything. :)

File: tests/lib/rules/eol-last.js [hunk 7]
```diff
@@ -26,10 +27,22 @@ ruleTester.run("eol-last", rule, {
         "var a = 123;\n\n",
         "var a = 123;\n   \n",
 
-        "\r\n",
```
Comment (kaicataldo) [inline tests/lib/rules/eol-last.js]: I think we should leave these tests - even though we we're not enforcing a Windows linebreak style any longer, we want to make sure that this rule is still compatible with these cases

File: tests/lib/rules/eol-last.js [hunk 8]
```diff
@@ -26,10 +27,22 @@ ruleTester.run("eol-last", rule, {
...
-        "var a = 123;\r\n",
-        "var a = 123;\r\n\r\n",
-        "var a = 123;\r\n   \r\n"
+        { code: "", options: ["unix"] },
```
Comment (kaicataldo) [inline tests/lib/rules/eol-last.js]: Can we group together the tests that are for the deprecated options with a comment labeling them as such in both the valid and invalid tests (maybe at the end of each grouping)? Will make it much easier for the next person who might be removing the options in a future major version.

File: docs/rules/eol-last.md [hunk 9]
```diff
@@ -1,4 +1,4 @@
-# Require file to end with single newline (eol-last)
+# Require or disallow file to end with single newline (eol-last)
```
Comment (kaicataldo) [inline docs/rules/eol-last.md]: I believe we are making these titles start with a lowercase character now - can someone else on the team  confirm?

File: lib/rules/eol-last.js [hunk 10]
```diff
@@ -1,62 +1,95 @@
...
+
+//------------------------------------------------------------------------------
 // Rule Definition
 //------------------------------------------------------------------------------
 
 module.exports = {
     meta: {
         docs: {
-            description: "enforce at least one newline at the end of files",
+            description: "enforce at least one newline (or absence thereof) at the end of files",
```
Comment (kaicataldo) [inline lib/rules/eol-last.js]: Can we make this consistent with the rule description in the title of the docs file? Thoughts on "require or disallow newlines at the end of files"?

**Comment by kaicataldo** [comment]
Just a few small comments

File: lib/rules/eol-last.js [hunk 11]
```diff
@@ -1,62 +1,95 @@
 /**
- * @fileoverview Require file to end with single newline.
+ * @fileoverview Require or disallow file to end with single newline.
  * @author Nodeca Team <https://github.com/nodeca>
+ * @author kdex <https://github.com/kdex>
```
Comment (kaicataldo) [inline lib/rules/eol-last.js]: We reserve the `@author` field for the file creator - your contribution will be in the Git history though!

File: tests/lib/rules/eol-last.js [hunk 12]
```diff
@@ -1,6 +1,7 @@
 /**
- * @fileoverview Check, that file is ended with newline, and there are no multiple empty lines at the end.
+ * @fileoverview Tests for eol-last rule.
  * @author Nodeca Team <https://github.com/nodeca>
+ * @author kdex <https://github.com/kdex>
```
Comment (kaicataldo) [inline tests/lib/rules/eol-last.js]: Same as https://github.com/eslint/eslint/pull/6952/files#r78205009

File: lib/rules/eol-last.js (hunk 11 above)
Comment (kdex) [inline lib/rules/eol-last.js]: Sorry! I'll be removing mine. Note that I have seen cases where additional authors were added [years](https://github.com/eslint/eslint/blame/4126f12f792231c844859e13a3fbd6b7ff78c4ee/lib/rules/no-plusplus.js) after the file has been created, up to [this year](https://github.com/eslint/eslint/blame/c67632259783c9630831bce5f272eb6160ee086c/lib/rules/no-duplicate-case.js), which I followed. Did these cases slip under the radar?

Comment (platinumazure) [inline lib/rules/eol-last.js]: @kdex Yes, probably. Maybe we'll sweep through those later. (Some cases are legitimate collaborations at time of file creation, though.) Thanks for your understanding!

**Comment by kdex** [comment]
@kaicataldo The issues from above should be fixed. One last thing: Is `@fileoverview` supposed to be consistent with either of `module.exports.meta.docs.description` or the description in the title of `eol-last.md`?

File: lib/rules/eol-last.js (hunk 11 above)
Comment (kaicataldo) [inline lib/rules/eol-last.js]: No need to apologize! Looks like some cases where we missed it, and should probably clean them up at some point.

**Comment by kaicataldo** [comment]
@kdex I think it makes sense for them to be, yes, though those seem to be all over the place in the codebase. Would you mind making them consistent?

**Comment by kaicataldo** [comment]
Other than one tiny comment, LGTM. Waiting for a few others to review before merging. Thanks for being so responsive and for doing great work on this, @kdex! :+1:

**Comment by kdex** [comment]
@kaicataldo This commit should make them more consistent, but having a look into the codebase, they're bound to be inconsistent anyway, as `@fileoverview`s start with capital letters and sometimes end with periods, whereas `meta.description` and documentation titles don't.

Having these two match implies that `@fileoverview` comments should also start in lowercase letters, which currently not the case.

If there are no objections, I would make all `@fileoverview` comments consistent for the entire codebase in another PR, making them start with lowercase letters and removing the periods.

Anyway, this PR should be ready-to-merge now, please review. :smile:

**Comment by kaicataldo** [comment]
LGTM! Just want to make sure we get some other eyes on this before merging. Thanks again for contributing, @kdex!

Having the `@fileoverview` comments be consistent across all the rules seems like a nice thing to me - if that's something you'd like to do and you want to make an issue we can see what the rest of the team thinks.

**Comment by kaicataldo** [comment]
@nzakas @vitorbal @mysticatea @platinumazure Would love at least one more pair of eyes on this before I merge :)

File: lib/rules/eol-last.js [hunk 13]
```diff
@@ -1,62 +1,94 @@
...
-
             Program: function checkBadEOF(node) {
-
                 const sourceCode = context.getSourceCode(),
                     src = sourceCode.getText(),
-                    location = {column: 1},
-                    linebreakStyle = context.options[0] || "unix",
-                    linebreak = linebreakStyle === "unix" ? "\n" : "\r\n";
+                    location = {
+                        column: 1,
```
Comment (platinumazure) [inline lib/rules/eol-last.js]: Should this be zero-based? I thought I saw this come up elsewhere that columns should be zero-based on report and will automatically be converted to one-based before output.

**Comment by platinumazure** [comment]
Just one question about the report column, otherwise LGTM.

File: tests/lib/rules/eol-last.js [hunk 14]
```diff
@@ -44,6 +63,52 @@ ruleTester.run("eol-last", rule, {
...
+            options: ["never"],
+            errors: [{ message: "Newline not allowed at end of file.", type: "Program" }],
+            output: "var a = 123;"
+        },
+        {
+            code: "var a = 123;\r\n\r\n",
+            options: ["never"],
+            errors: [{ message: "Newline not allowed at end of file.", type: "Program" }],
+            output: "var a = 123;"
+        },
```
Comment (vitorbal) [inline tests/lib/rules/eol-last.js]: Could we add a test for two unix-style newlines being fixed to no newlines?

**Comment by vitorbal** [comment]
This looks really good, just suggested one additional test case. Great work, @kdex!

**Comment by kdex** [comment]
Thanks for reviewing, everyone!

@vitorbal Test added! Thank you.

@platinumazure That's quite a good catch, since neither assuming the column to be `0` nor `1` really makes sense for this rule. Come to think about it, the last EOL is really supposed to be the last character in a file, so it should be reported for: last line, last column, and **not**: last line, first column.

Another option would be to report for column `0`, but then I'd be forced to report the warning for a line number that doesn't even exist in the file (last line + `1`). Not sure if all editors can integrate that properly.

Making this change requires me to fix three tests in `cli-engine.js` and `eslint.js`.

Can I get some feedback about the expected semantics before I commit?

File: lib/rules/eol-last.js (hunk 13 above)
Comment (kaicataldo) [inline lib/rules/eol-last.js]: Good catch, and yeah, `line` is one-based and `column` is zero-based: http://eslint.org/docs/developer-guide/working-with-rules#contextreport

**Comment by kaicataldo** [comment]
Last line, last column makes sense to me, since that's always the location this rule checks!

**Comment by kaicataldo** [comment]
Actually, I'd be okay with us landing this as is (since the current reported column is the pre-existing behavior and not something you added), since you've done so much good work on it and I know there's been a lot of back and forth. Would hate to hold this up any more. We can always open another bug fix issue for the location if it is indeed incorrect.

**Comment by kdex** [comment]
Agreed! Here's the test @vitorbal mentioned.

I already fixed the column issue in my local version; will commit and open a new issue/PR for that once we get this one merged. :smile:

**Comment by kaicataldo** [comment]
Having the reported location be the last character of the file makes sense to me - let's see what the rest of the team has to say :)

This looks really great! Thanks again @kdex for your patience with the back and forth and for being so diligent with this PR - I think we ended up finding a much better solution through the process. Thanks!

**Comment by vitorbal** [comment]
Thanks for your contribution, @kdex! :tada: :tada:
